### ISSUE TITLE
History is lost

### ISSUE THREAD

**Issue by FrenchBen** [warpdotdev/Warp#5212]
### Dupe Check

- [X] I have [searched Warp bugs](https://github.com/warpdotdev/warp/issues?q=is%3Aissue+label%3ABugs) and there are no duplicates

### Describe the bug

Despite history files existing, several shell env changes, history in Warp isn't retained.
The question then is: Where is the actual history preference set within Warp?

### To reproduce

1. Use wap
2. Execute several "similar" commands like `curl ...`
3. Expect warp to recall older curl commands

### Expected behavior

Warp should remember everything since the beginning of time :)

### Operating system

MacOS

### Operating system and version

14.5

### Shell Version

zsh 5.9

### Current Warp version

v0.2024.07.16.08.02.stable_03

### Regression

No, this bug or issue has existed throughout my experience using Warp

### Does this block you from using Warp daily?

No

### Is this an issue only in Warp?

Yes, I confirmed that this only happens in Warp, not other terminals.

### Warp Internal (ignore): linear-label:b9d78064-c89e-4973-b153-5178a31ee54e

None

**Comment by FrenchBen** [comment]
Not only is history from Zsh not respected, the history written to the HISTFILE isn't shared across tabs or pane, rendering cross-tab functionality painful.

### Replicate bug

1. Create 2 tabs (or pane)
2. Type something
3. Try to get the history of what you typed from another pane


Histfile will show the command:
![Screenshot 2024-08-14 at 10 28 51](https://github.com/user-attachments/assets/71c60ec3-e012-40cd-8458-6efe0db7d711)

yet Warp cannot "find it"
![Screenshot 2024-08-14 at 10 28 39](https://github.com/user-attachments/assets/21160e76-e714-48c9-9026-cfd5574c50c0)

![Screenshot 2024-08-14 at 10 28 26](https://github.com/user-attachments/assets/09ba9f31-3708-4d0f-ae3c-0cc620140a29)

**Comment by dannyneira** [comment]
Hi @FrenchBen sorry to hear about this. Warp doesn't share history immediately across sessions, although this setting may be enabled in your zsh config. Currently, Warp does read the HISTFILE upon loading a new session, so any commands you ran in other sessions should show up when you open a new window/tab/pane.

**Comment by FrenchBen** [comment]
@dannyneira thank you for your note - How does Warp save its history if not written to the HISTFILE directly, which would make it cross-session?
Is this an issue of the history being written, but Warp not re-loading it?

**Comment by dannyneira** [comment]
> How does Warp save its history if not written to the HISTFILE directly, which would make it cross-session?

 Yes they are written to the HISTFILE as with normal terminals, but the history doesn't get consolidated across existing sessions (regardless of rc file settings), they only get combined when a new session is made.

> Is this an issue of the history being written, but Warp not re-loading it?

Warp does reload the history and consolidate it, on new sessions only. I believe there is a feature request for option to consolidate on current sessions, or respect rc file settings for consolidation of history, but this isn't an option yet.

**Comment by FrenchBen** [comment]
@dannyneira I noticed a few `warp_` shell functions. Would a reload shell script be the right place to do this (aka launch `warp_hist` or similar)?
How would I go about creating a feature design discussion? Can you point me to the code that loads / consolidates history on new session?

**Comment by dannyneira** [comment]
Not sure of which code exactly does this, but starting a new session will consolidate the history as far as I'm aware you can't workaround or disable this behavior

**Comment by tzaffi** [comment]
Does anyone have a workaround? For me this is borderline blocking me from using Warp.

**Comment by francescoPadovani81** [comment]
Hi,
also, when I run the _history_ command in warp it only shows me the last 500 commands. It seems not to respect the variable HISTSIZE (which in my case e.g. is 10K). This limits my warp usage: every now and then it happens that I have to combine the history + grep + diff command and I need to have the full history available. This force me to use another terminal.
The Command Search function (Ctrl+r) of warp on the other hand is perfectly able to search the entire history file, so I don't understand why the history command is limited instead.

Sorry, I assumed that my problem was related to this issue. If I'm wrong tell me and I'll open a new one (I couldn't find a specific one with my problem).

Thank you
